### ISSUE TITLE
Frontier: First Encounters (1995) fails to run on non-x86 platforms

### ISSUE THREAD

**Issue by kcgen** [dosbox-staging/dosbox-staging#2900]
### Context

Currently the eXoDOS project runs this game using [the DOSBox Daum fork](https://github.com/ykhwong/dosbox-svn-daum), which is understood to have specific emulation fixes not currently in upstream DOSBox (0.74.x and SVN) or active forks such as X, Pure, and Staging.

DOSBox Daum had 51 releases that included DOSBox SVN's commits ~r3600 through ~r3900, [spanning 2010 through 2015](http://ykhwong.x-y.net/).

### The goal of this ticket is to answer the following

Is Daum _still_ needed for this game?
- [ ] Does DOSBox SVN ([download latest](http://www.emucr.com/2023/09/dosbox-svn-r4483.html)) play it OK?
- [ ] Does DOSBox X ([download latest](https://dosbox-x.com/)) play it OK?
- [ ] Does DOSBox Staging ([download latest](https://dosbox-staging.github.io/downloads/development-builds/)) play it OK?

If Daum plays it better:
- [ ] What aspect is better?
- [ ] Which Daum release was the **_first_** to fix it?
    Use these builds [here](http://ykhwong.x-y.net/subpages/old.php) to narrow it down.

### How you can help

 1. Download the above lastest DOSBox SVN, X, and Staging builds.
 1. Ensure you have [eXoDOS v5](https://www.retro-exo.com/exodos.html) (or eventually v6)
 1. Find this game inside the eXoDOS installed folders (look for: `frontier/`)
 1. Run the game using the DOSBox SVN, X, and Staging builds.
    _If it plays fine (everything looks, sounds, and handles OK) then simply indicate as such. Your task is done!_

If SVN, X, and Staging all have problems, then try it with Daum:
 1. Download the [first and last Daum releases](http://ykhwong.x-y.net/subpages/old.php).
    Note: _Daum had the most releases for Microsoft Windows, so having a Windows system (or VM) will help the most to answer the last question above._
 1. Does the latest Daum release play it correctly? If so: summarize/describe how it's better than DOSBox SVN, X, and Staging (screenshots, video, audio records all help too).
 1. Find the earliest (or oldest) Daum release that fixed it.
    1. Run the Daum release half way between the known good release (where the game playing correctly) and bad release (where the game doesn't playing correctly) until you've narrowed it down to adjacent releases.
    1. It will take _at-most_ seven runs of different Daum versions to narrow this down (using this middle-selecting technique); so it's not as daunting as it seems! :-)

---

### Follow up by code contributors

If Daum plays it better, your answers (especially the last one that narrows down the fix to a specific release) are critical to helping isolate the chunk of code fixed the given bug.

Pull requests should credit:
- @ykhwong as the original code author,
- the test team member who confirmed and isolated the Daum release,
- yourself (PR author) as integrating the fix

The PR should include the `imported-from:` URL pointing to DOSBox Daum. For an example, see commit a35c7c889.

**Comment by Grounded0** [comment]
Will be reviewed soon.

**Comment by Grounded0** [comment]
We already know this game doesn't run on macOS/ARM64 so it should be checked with Linux/ARM64 does it run or not.

**Comment by Grounded0** [comment]
Connected issues and PRs:

#1071

**Comment by Grounded0** [comment]
Drawing glitch reported in #1071 has been corrected when using `machine=svga_s3`:
<img width="1392" alt="Screenshot 2023-09-20 at 6 11 06" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/cacb98ab-8de7-479d-9a84-16d184082c02">

**Comment by Grounded0** [comment]
I'll give this a good 30 minute play when i wake up to close this.

**Comment by Grounded0** [comment]
After playing for 35 minutes i can't find issues with the game. Closing. @kklobe can comment is this his view also as a regular FFE player.

**Comment by Burrito78** [comment]
Results (macOS Ventura 13.5.2/Apple M2):

**DOSBox Staging 0.81.0-alpha-1689**
When i press space in the intro it starts again, pressing space one more time crashes DOSBox.

**DOSBox SVN	snapshot 12.09.23**
Very slow in the intro already. When i press space in the intro it starts again, pressing space one more time crashes DOSBox.

**DOSBox-X v2023.09.01**
All OK

I don't run Windows so i can't test with Daum.

@Grounded0 Game shouldn't crash Staging, no matter what.

**Comment by Grounded0** [comment]
Can't get it to crash either with 0.80.1 or 0.81 dev. I installed mine from archive.org FFE Enhanced CD version and installed latest 1.06 patch.

Also spent by now 45 minutes playing the game with no issues. This is an AAA title so i spent extra energy to get these right. @johnnovak should we close this or not?

**Comment by johnnovak** [comment]
@Grounded0 I think the crash is a bit worrying. Could you check with @Burrito78 to make sure you guys are testing the same version? And/or document which version doesn't crash, and which does? That would be useful, because a bug is a bug, and yeah, nothing should outright crash Staging.

**Comment by Grounded0** [comment]
Running clean copy from archive.org patched to release 1.06. Can't replicate that crash. You'll guys try to figure this out since i've already spent enough playtime to prove it runs right.

**Comment by Grounded0** [comment]
@johnnovak

Question is does issue with macOS/ARM64 make the game unplayable or not. Your call. 😁

**Comment by johnnovak** [comment]
> @johnnovak
> 
> Question is does the issue with macOS/ARM64 make the game unplayable or not. Your call. 😁
> 
> @kklobe spent a lot of time and energy in order to fix this but couldn't.

Well, I think let's move on, close this one off, and create a ticket for it in the backlog (if we don't have one already). Just to keep track of it, but it's kinda in the "too hard to fix for ARM" bucket then.

**Comment by Grounded0** [comment]
@johnnovak

You can close this to execute better authority so this doesn't become open/close/open.

**Comment by Grounded0** [comment]
This macOS/ARM64 issue should be checked with Linux/ARM64 does it happen there or not as i stated earlier.

**Comment by Burrito78** [comment]
I've used the eXoDOS version which is 1.05.
I can confirm that using DOSBox Staging DEV with Rosetta, the game runs fine.

I will create a backlog ticket later and close this here again. 

Thanks guys!

**Comment by classilla** [comment]
I don't mind checking but I can't see where the exact binary people are using is. I'm not familiar with this title.

**Comment by kcgen** [comment]
Unfortunately it's a CDROM-based game so too large to attach a reproduction pack, however Archive.org has the game preserved:

https://archive.org/details/firstencounters1995usrelease

**Comment by kklobe** [comment]
This is a small version I use for testing on my Mac, crashes after two spaces as per @Burrito78 

[firstenc.zip](https://github.com/dosbox-staging/dosbox-staging/files/12689202/firstenc.zip)

**Comment by kklobe** [comment]
FWIW, this one still annoys me, since it makes no sense to me that the pure C++ code in `paging.cpp` is somehow compiled differently enough between x86-64 and arm64 to cause one to work and the other not to.

**Comment by kcgen** [comment]
@kklobe , @ebenbruyns --

Some very tantalizing feedback from ARM64 + circa-2015 Daum:

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/a83770a4-51e4-4856-8a6b-242af6333ffe

SVN crashes just like Staging (see bottom of this message), so the fix is isolated in Daum!

@kklobe, here's my modified source tree based on @ebenbruyns's  reconstructed SCM for Daum: 

[:file_folder: junkrunner-dosbox-daum-coarse-scm-reconstruction.tar.gz](https://github.com/dosbox-staging/dosbox-staging/files/12689398/junkrunner-dosbox-daum-coarse-scm-reconstruction.tar.gz)

I used `brew install sdl12-compat sdl_net`, and then `./autogen.sh`, `./configure`, `make -j12` should produce the binary.

@ebenbruyns has separated out Daum's specific changes from SVN and X, here:

https://github.com/dosbox-staging/dosbox-staging/compare/main...ebenbruyns:dosbox-staging:daum-2015

Specifically in this commit:

![2023-09-21_10-07](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/084978df-1af9-4416-9bab-b9c7705dd146)

You can ignore all the POD save-state stuff, and see the essence of the emulation changes - include the page fault handling changes, which we can see in action when the intro switches to the in-game logic.

I'm still not sure why x86-64 is avoiding this page fault.. so it could be that a bug still exists, however Daum's Page Fault handling mitigates it?

---

For reference, here's a clean build and run of SVN on ARM64:

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/b3149e92-8ea4-4d87-9a1a-5f3519b26be6

**Comment by classilla** [comment]
Yes, it also crashes on `ppc64le`.

```
Stack trace:
30      0x7fff8954846c __libc_start_main + 428
29      0x7fff8954822c /lib64/libc.so.6(+0x3822c) [0x7fff8954822c]
28          0x100c6308 main + 24
27          0x1032cf7c sdl_main(int, char**) + 2316
26          0x100e3b24 Config::StartUp() + 36
25          0x105091d8 SHELL_Init() + 4424
24          0x10506b68 DOS_Shell::Run() + 712
23          0x10504ba0 DOS_Shell::RunBatchFile() + 224
22          0x10504234 DOS_Shell::ParseLine(char*) + 292
21          0x10513844 DOS_Shell::DoCommand(char*) + 404
20          0x10525070 DOS_Shell::ExecuteProgram(std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >) + 2544
19          0x1016555c CALLBACK_RunRealInt(unsigned char) + 92
18          0x100d0dc0 DOSBOX_RunMachine() + 48
17          0x100d0c64 /home/spectre/src/dbs-ppc/build/release/dosbox() [0x100d0c64]
16          0x101b3e44 CPU_Core_Normal_Run() + 42420
15          0x10168708 CPU_Push32(unsigned long) + 104
14          0x103bd010 mem_writed(unsigned int, unsigned int) + 128
13          0x10177c50 InitPageHandler::writed(unsigned int, unsigned int) + 1184
12          0x10174358 PAGING_PageFault(unsigned int, unsigned int, unsigned int) + 216
11          0x100d0dc0 DOSBOX_RunMachine() + 48
10          0x100d0c64 /home/spectre/src/dbs-ppc/build/release/dosbox() [0x100d0c64]
9           0x10174028 /home/spectre/src/dbs-ppc/build/release/dosbox() [0x10174028]
8           0x1019c440 CPU_Core_Full_Run() + 2464
7           0x1017851c InitPageHandler::readb(unsigned int) + 652
6           0x10174358 PAGING_PageFault(unsigned int, unsigned int, unsigned int) + 216
5           0x100d0dc0 DOSBOX_RunMachine() + 48
4           0x100d0c64 /home/spectre/src/dbs-ppc/build/release/dosbox() [0x100d0c64]
3           0x10174028 /home/spectre/src/dbs-ppc/build/release/dosbox() [0x10174028]
2           0x101a30b4 CPU_Core_Full_Run() + 30228
1           0x1017808c InitPageHandler::readd(unsigned int) + 700
0           0x100f7a8c E_Exit(char const*, ...) + 124
2023-09-21 10:27:34.589 | ABORT: Pagefault didn't correct page
Abort (core dumped)
```

I'll play around a bit with it. Thank you for the test vector, @kklobe .

**Comment by weirddan455** [comment]
@classilla Can you try this patch you suggested of changing `uint32_t` to `uint64_t`?  If this is indeed the same bug, it would be good to know.

https://github.com/dosbox-staging/dosbox-staging/pull/2863#issuecomment-1717003839

**Comment by Grounded0** [comment]
@Burrito78 

I think we'll try to solve this issue on this open ticket for now. I'll add this to problematic games on the megaticket.

**Comment by kcgen** [comment]
> @Burrito78
> 
> I think we'll try to solve this issue on this open ticket for now. I'll add this to problematic games on the megaticket.

Good call on this new information, @Grounded0.

It's indeed a case where Daum runs it and SVN (and by extension Staging) does not, now reproduced on two 64-bit non-x86 architectures.

**Comment by classilla** [comment]
> @classilla Can you try this patch you suggested of changing `uint32_t` to `uint64_t`? If this is indeed the same bug, it would be good to know.
> 
> [#2863 (comment)](https://github.com/dosbox-staging/dosbox-staging/pull/2863#issuecomment-1717003839)

Interesting thought, but no, doesn't help. There's actually an exception triggered *within* the protected mode handler before DOSBox goes down.

![boom](https://github.com/dosbox-staging/dosbox-staging/assets/12717725/a31be58e-1ec5-4848-b6d4-24c8ab663a81)

**Comment by classilla** [comment]
In a debug build, it doesn't crash, it just goes into a never-ending loop of illegal instruction/illegal address exceptions, mostly mangled.

**Comment by classilla** [comment]
I need to cobble that together. In the meantime, this does at least wallpaper the crash and you can restart DOSBox, but the game is still a write-off.

```
diff --git a/src/cpu/paging.cpp b/src/cpu/paging.cpp
index 3121225ae..844f77af7 100644
--- a/src/cpu/paging.cpp
+++ b/src/cpu/paging.cpp
@@ -128,17 +128,19 @@ static struct {
 static Bits PageFaultCore()
 {
        CPU_CycleLeft+=CPU_Cycles;
        CPU_Cycles=1;
        Bits ret=CPU_Core_Full_Run();
        CPU_CycleLeft+=CPU_Cycles;
        if (ret<0) E_Exit("Got a dosbox close machine in pagefault core?");
        if (ret) 
+{ if (ret==1) { fprintf(stderr, "giggity\n"); PAGING_ClearTLB(); } else
                return ret;
+}
        if (!pf_queue.used) E_Exit("PF Core without PF");
        PF_Entry * entry=&pf_queue.entries[pf_queue.used-1];
        X86PageEntry pentry;
        pentry.load=phys_readd(entry->page_addr);
        if (pentry.block.p && entry->cs == SegValue(cs) && entry->eip==reg_eip) {
                cpu.mpl=entry->mpl;
                return -1;
        }
```

**Comment by classilla** [comment]
Actually, the `PAGING_ClearTLB()` isn't needed. It looks like all it's doing is holding it in the PageFaultCore until the emulator can be rescued. Hrm.

**Comment by classilla** [comment]
In Daum's `src/hardware/memory.cpp`, `gcc` doesn't like this. What's supposed to be there?

```
//save state support
extern void* VGA_PageHandler_Func[16];

Bit32u Memory_PageHandler_table[] =
{
#if 0
        (Bit32u) NULL,
        (Bit32u) &ram_page_handler,
        (Bit32u) &rom_page_handler,

        (Bit32u) VGA_PageHandler_Func[0],
        (Bit32u) VGA_PageHandler_Func[1],
        (Bit32u) VGA_PageHandler_Func[2],
        (Bit32u) VGA_PageHandler_Func[3],
        (Bit32u) VGA_PageHandler_Func[4],
        (Bit32u) VGA_PageHandler_Func[5],
        (Bit32u) VGA_PageHandler_Func[6],
        (Bit32u) VGA_PageHandler_Func[7],
        (Bit32u) VGA_PageHandler_Func[8],
        (Bit32u) VGA_PageHandler_Func[9],
        (Bit32u) VGA_PageHandler_Func[10],
        (Bit32u) VGA_PageHandler_Func[11],
        (Bit32u) VGA_PageHandler_Func[12],
        (Bit32u) VGA_PageHandler_Func[13],
        (Bit32u) VGA_PageHandler_Func[14],
        (Bit32u) VGA_PageHandler_Func[15],
#endif
};
```

**Comment by kcgen** [comment]
Sorry @classilla ; I should have done a better job of knocking out these problematic save state serialization objects. 

The `Memory_PageHandler_table` is only used in Daum's save/restore state code, so no harm in it being an empty placeholder (whatever pacifier your compiler). 

I made these changes for clang on ARM64, so looks like gcc handles it a bit differently.

**Comment by classilla** [comment]
Okay, I coerced it to compile, and the game seems to run normally on this `ppc64le` workstation, so the Daum code does appear to solve it universally.

**Comment by johnnovak** [comment]
> Okay, I coerced it to compile, and the game seems to run normally on this `ppc64le` workstation, so the Daum code does appear to solve it universally.

The first major improvement we found in Daum AFAIK. Now hopefully it won't cause regressions in other games.

**Comment by ebenbruyns** [comment]
Did all my hard slaving away at that codebase contribute to this? I'm deperate for a win here...

**Comment by kklobe** [comment]
This issue is like a splinter in my mind.

I'm going to cut a branch that will do extensive loguru logging to a file of all paging (and maybe more) operations, and see if I can diff the x86-64 and arm64 builds on my Mac. 

Will report back.

**Comment by kcgen** [comment]
Good look @kklobe ; logic is on you're side 👍

There will be no more corners of the Venn diagram when you start painting code good and bad with these comparisons!

**Comment by rderooy** [comment]
Some additional testing from my side.
https://www.mobygames.com/game/351/frontier-first-encounters/
https://www.freegameempire.com/games/Frontier-First-Encounters/manual-pdf

Not sure why this has both `aspect=fase` and "ASPECT FALSE" in the autoexec section (the later of which will be ignored by default on DOSBox-X).

Gives 3 sound options on startup. Testing was done with SB.

## DOSBox 0.74-3
Some visual glitches during the intro.  Pressing space in the intro a few times results in lots of HEX being printed to the screen and ultimately a DOSBox crash: `terminate called after throwing an instance of 'char*'`

## DOSBox ECE r4230 - as included with eXoDOS
Does not crash, but similar visual glitches to DOSBox 0.74-3 in both the intro and during gameplay.

## DOSBox Daum - as included with eXoDOS
Visual glitches during intro resolved, seems to work fine in limited testing

## DOSBox Staging 0.80.1
Same as Daum

## DOSBox-X 2023.09.01
Same as Daum

# eXoDOS dosbox.conf
```ini
# Config file for Dosbox-ECE r4230
[sdl]
fullscreen=false
fulldouble=false
fullresolution=0x0
windowresolution=1280x960
output=direct3d
mapperfile=mapper.map
[dosbox]
machine=svga_s3
memsize=32
[render]
aspect=false
scaler=normal2x
[cpu]
core=auto
cputype=auto
cycles=30000
[pci]
[mixer]
[midi]
mididevice=mt32
midiconfig=.\mt32\SoundCanvas.sf2
mt32.romdir=.\mt32
[sblaster]
[gus]
[speaker]
[joystick]
[serial]
[dos]
[ipx]
[autoexec]
@mount c .\eXoDOS\
imgmount d .\eXoDOS\frontier\cd\ELITE3_FIRST_ENCOUNTERS.ISO -t cdrom
@c:
cls
ASPECT FALSE
@cd frontier
@call run
exit
```
# eXoDOS run.bat
```dos
:menu
@echo off
cls
echo.
echo Press 1 for Frontier: First Encounters w/ SoundBlaster
echo Press 2 for Frontier: First Encounters w/ MT-32
echo Press 3 for Frontier: First Encounters w/ Sound Canvas
echo Press 4 to Quit
echo.
choice /C:1234 /N Please Choose:

if errorlevel = 4 goto quit
if errorlevel = 3 goto SC55
if errorlevel = 2 goto MT32
if errorlevel = 1 goto SB16

:SB16
CONFIG -set "mididevice=default"
copy .\sb16\*.* .\
cls
@FIRSTENC
goto quit

:MT32
CONFIG -set "mididevice=mt32"
copy .\mt32\*.* .\
cls
@FIRSTENC
goto quit

:SC55
CONFIG -set "mididevice=synth"
copy .\sc55\*.* .\
cls
@FIRSTENC
goto quit

:quit
```

**Comment by Grounded0** [comment]
Connected issues and PRs:

- #2404

**Comment by Grounded0** [comment]
I've been thinking that since x87 is available in x86-64 long mode:

![amd64digest](https://github.com/dosbox-staging/dosbox-staging/assets/48192273/243f7df0-aeb6-4fe1-8b64-15f25d86eafd)

That this issue could be related to floating point precision since FFE has a complex orbital mechanics simulation and may use 80-bit extended precision to get that math right. ARM64 and Power ISA would naturally round off these to 64-bit accuracy instead of IEEE 754 extended precision.

**Comment by Grounded0** [comment]
You can check how IEEE 754 extended precision conformant you are by running this tool:

[MCPDIAG.ZIP](https://github.com/dosbox-staging/dosbox-staging/files/12748143/MCPDIAG.ZIP)

My results on x86-64:

<img width="1392" alt="Screenshot 2023-09-28 at 12 08 05" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/08bc3e9c-f298-4650-8c0a-9237b84b2a46">

**Comment by kklobe** [comment]
I am able to reproduce the issue with the x86-64 build with normal core on my M1 Mac, so there seems to be something that the dynx86 core is compensating for, but I haven't had the time to dive deep.

**Comment by johnnovak** [comment]
I'm afraid it wouldn't be trivial to support 80-bit floating point precision @Grounded0. But yeah, it would be nice to support it natively. And it's not supported on ARM, it's an Intel-special thing.

**Comment by kklobe** [comment]
I'm also not even sure most modern compilers support that for x86-64, I think most (all?) of them use SSE/AVX for floating point?

**Comment by Grounded0** [comment]
@GranMinigun on IRC: Looking at src/fpu/fpu_instructions_x86.h, it directly calls native FADD and other instructions.

**Comment by kklobe** [comment]
> @GranMinigun on IRC:
> 
> > Looking at `src/fpu/fpu_instructions_x86.h`, it directly calls native `FADD` and other instructions.

Ok, I'm going to give my branch a try with dynx86 but with `C_FPU_X86` disabled so it bypasses that file.

**Comment by johnnovak** [comment]
Yeah, so it's a bit shit that x86-64 supports 80-bit precision, while ARM doesn't... Even if we could do that easily (and I don't think we can), that's too much cross-platform discrepancy for my taste ("look bro, you _lose precision_ when you run the same program on a new Mac versus your old Windows box... yeah, it's completely normal, moving on" 🥶)

**Comment by Burrito78** [comment]
![Bildschirmfoto 2023-09-28 um 14 48 04](https://github.com/dosbox-staging/dosbox-staging/assets/61995128/217a156c-c526-4aca-862b-99f7d77c5797)

**Comment by Grounded0** [comment]
@Burrito78

Run one for us with Rosetta just for comparison. Thanks.

**Comment by Burrito78** [comment]
> @Burrito78
> 
> Run one for us with Rosetta for comparison. Thanks.

![Bildschirmfoto 2023-09-28 um 15 11 06](https://github.com/dosbox-staging/dosbox-staging/assets/61995128/4ee08be0-b145-4b5f-92bc-5698648bd916)

**Comment by Grounded0** [comment]
Rosetta is more accurate emulating x86 than Intel is on their own silicon. 😁

**Comment by weirddan455** [comment]
>  I'm also not even sure most modern compilers support that for x86-64, I think most (all?) of them use SSE/AVX for floating point?

It's part of the x86_64 ABI to use the SSE xmm* registers for floating point.  When you pass a float or a double to a function, it gets passed in those registers.  I think it's also faster than the old x87 instructions even if you're only using the scalar non-wide instructions (ex. `mulss` vs `mulps`).  I don't know if there's a way to "force' the compiler to use the 80 bit x87 instructions but it doesn't by default.

>  Yeah, so it's a bit shit that x86-64 supports 80-bit precision, while ARM doesn't... Even if we could do that easily (and I don't think we can, I think MSVC doesn't support 80-bit FPU mode), that's too much cross-platform discrepancy for my taste ("Look bro, you lose precision when you run the same program on a new Mac versus your old Windows box... Oh, but if you have an older Intel Mac, it's fine again! Yeah, it's completely normal, moving on" 🥶)

Ideally, we should be emulating 80 bit precision on platforms that don't support it natively.  I mean we are an emulator after all, right?  This might be an area where someone decided to take the simpler and more performant option over the more accurate option since a lot of DOS games use integer only arithmetic.

In the early days of the PC, floating point was only available on a math co-processor that you had to buy and install separately.  Plus, even if you had one, they tended to be slower than integer operations so a lot of games decided to just use integer or write their own fixed point solution if they needed to.

**Comment by Grounded0** [comment]
I managed to get a perfect score by running this with `core=dynamic` on x86-64:

<img width="1392" alt="Screenshot 2023-09-28 at 16 45 27" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/274127f4-451f-48fd-9c2c-276d39db5803">

**Comment by johnnovak** [comment]
> Ideally, we should be emulating 80 bit precision on platforms that don't support it natively. I mean we are an emulator after all, right? 

Yes, unless it's dog slow. Like cycle accurate CPU emulation with prefetching, pipelining, etc. On current hardware, you can't even emulate a 286 at that accuracy, from memory, so it's rather useless and an overkill. I have a hunch software FPU emulation would be similar.

> Plus, even if you had one, they tended to be slower than integer operations so a lot of games decided to just use integer or write their own fixed point solution if they needed to.

Yeah, I used 16.16 fixed point in 32-bit regs and precalculated sine tables, etc. That was pretty standard in demo coding circles, so in games too. We used the FPU later for runtime precalc only, that's it. I'm pretty sure nothing used the FPU in anger pre-Quake, except for the odd flight sim. Maybe Falcon too? And these later DOS era games are rather demanding, so performance might go to shit with a soft FPU approach.

Hence my view: big effort, low payoff, potentially. But then, someone can try I see I guess.

Also, I'm not sure 128-bit floats would work. What about games that mess around with FPU values at the bit level? You can do a lot of bit twiddling tricks on floats. Actually, how can those work even now?

**Comment by kklobe** [comment]
I did a quick test of [MPFR](https://www.mpfr.org) and this little loop took ~550ms:

```cpp
    for (int i = 0; i < 1000000; ++i) {
        mpfr_add(sum, a, b, MPFR_RNDN);
        mpfr_cos(cos, sum, MPFR_RNDN);
    }
```

I'm sure MPFR might not be the fastest possible implementation, but even that rough order of magnitude seems like a non-starter.

**Comment by johnnovak** [comment]
> I'm sure MPFR might not be the fastest possible implementation, but even that rough order of magnitude seems like a non-starter.

Yeah, so what I thought. Really bad tradeoff of performance vs precision.

**Comment by kcgen** [comment]
That's pretty slow! 1000 calculation in half a millisecond. 
Assuming a 3 Ghz CPU.. that ~1,500,000 cycles per "add + cos.

**Comment by kcgen** [comment]
For First Encounter's crash/intro loop, recall the native macOS ARM64 build of Daum running it successfully (@classilla also confirmed it on ppcle64):

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/a83770a4-51e4-4856-8a6b-242af6333ffe

Given Daum didn't upgrade the FPU to 80-bits, I think this might be a red herring (just in this specific case) - and the fix probably lies in the `PAGING_NewPageFault()` logic on the [`forks/daum-scm-restoration`](https://github.com/dosbox-staging/dosbox-staging/compare/forks/daum-scm-restoration) branch.

**Comment by classilla** [comment]
That's my impression as well, that this hasn't anything to do with the FPU.

Power ISA does have a 128-bit long double which has the same size exponent as 80-bit long double on x86, so it should be just a matter of truncation. I do want to add this support since it would fix games like Carmageddon.

**Comment by kklobe** [comment]
Nope. The issue is with the paging, and there's something in the dynamic x86 core that compensates for it, I believe with the inlined memory access that isn't in the non-x86 dynamic core. I plan on digging more into this at some point.

**Comment by Grounded0** [comment]
Probably not an https://github.com/dosbox-staging/dosbox-staging/labels/FPU%20emulation issue but definitely a https://github.com/dosbox-staging/dosbox-staging/labels/non-x86 issue from what i've gathered.

**Comment by Grounded0** [comment]
Dropping https://github.com/dosbox-staging/dosbox-staging/labels/eXoDOS%2FWin3x since the game is not listed on eXoDOS megaticket:

- #3646

**Comment by Grounded0** [comment]
Workaround to run with macOS:

<img src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/85b293ee-eca5-40eb-b400-7599b3389d67" width="300">

**Comment by johnnovak** [comment]
@kklobe Wouldn't your recent FPU improvements help with this?

**Comment by kklobe** [comment]
> [@kklobe](https://github.com/kklobe) Wouldn't your recent FPU improvements help with this?

Not really, this one is a memory paging issue.
